### PR TITLE
WIP | adding draft of hallway documentation

### PR DESCRIPTION
**Pull request by pstjohn**
opening up a PR so I can add some comments

### CODE REVIEW

File: docs/source/hallway.rst [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+state, this endeavor becomes tractable.
+
+And so, we abandon the idea of “fixed” action spaces in favor of
+“parametric” action spaces. Here, **parametric** means that actions
+coincide with next states represented by feature vectors, rather than
+having a single, index-based interpretation. In other words, rather than
+actions as “move :math:`[ left, right ]`”, parametric actions can be
+thought of as “go to state :math:`[i, j]`” where the states :math:`i,j`
+each have a vector representation. Also unliked fixed spaces, the order
+doesn’t matter: we could equivalently say “go to state :math:`[j, i]`”.
```
Comment (pstjohn) [inline docs/source/hallway.rst:63]: Maybe not here, but somewhere in the docs (and likely in the corresponding paper) we'll want a *why graphenv?* section. Rllib does have some examples with parametric actions and masked actions, but we're (hopefully) making those concepts much easier for a user. (The rllib parametric actions demo just embeds (left, right) with static feature vectors for a cart-pole, for instance)

File: docs/source/hallway.rst [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+The **root** is the starting vertex of the graph search, here,
+:math:`i=0`.
+
+At each state in the search, a certain number of child states (or
+**children**) are accessible. In the figure above, we illustrate this
+using the color codes:
+
+-  black = current vertex
+-  white = child vertex
+-  gray = inaccessible, or **masked**, vertex
```
Comment (pstjohn) [inline docs/source/hallway.rst:94]: The masked vertex here is a bit odd -- it's not like we maintain a single global list of possible vertices and mask those that aren't reachable. For the molecule stuff (and probably other problems), we probably couldn't enumerate the whole graph *a priori* anyways.

Do you use the  notation later? Another concept we might want to discuss (perhaps not here) is that we're really handling some sort of acyclic Markov chain: the only inputs to a given node's value is it's observation, unlike the (state, action) embeddings that would translate here as (parent, child) observations.

Although we do have some support for intermediate rewards, right? So we're not completely history-agnostic when it comes to evaluating paths through the search tree.

File: docs/source/hallway.rst [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+.. code:: python
+
+   @property
+   def root(self) -> "HallwayState":
+       """
+       Returns:
+           HallwayState: initial state (vertex at index 0)
+       """
+       return self.new(0)
```
Comment (pstjohn) [inline docs/source/hallway.rst:228]: could probably skip these methods to get the the more illustrative `reward` and `children` methods.

File: docs/source/hallway.rst [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+           action_values = action_value_output(out)
+           action_weights = action_weight_output(out)
+
+           self.base_model = tf.keras.Model(
+               [cur_pos], [action_values, action_weights])
+
+HallwayEnv
+----------
+
+TODO: HERE
```
Comment (pstjohn) [inline docs/source/hallway.rst:362]: the only other concept I think we need to hit is the `max_num_actions` idea, where we're handling the padding / masking, but everything in `Vertex` is written from the point of view of a *single* instance.

File: docs/source/hallway.rst (hunk 1 above)
Comment (davebiagioni) [reply docs/source/hallway.rst:63]: i agree!  the scope for this notebook was more about the simple example, but i agree we need a motivation section -- perhaps just one click from the landing page.  

yeah, that's a good distinction, and i was struggling with how to express this.  we are different in that 

* actions <-> next states, rather than fixed actions conditioned on the current state (as in cartpole)
* the implementation is much simpler -- just define an obs space for a vertex and a model that is aware of it, the rest is done for the user

File: docs/source/hallway.rst (hunk 2 above)
Comment (davebiagioni) [reply docs/source/hallway.rst:94]: agreed, will stick with "inaccessible"

for the r_01 notation, yeah i was on the fence about this because from an implementation standpoint, we just attach the reward to a vertex.  but from the standpoint of the search, it is a collected during the transition from i->j -- will use this instead, esp since is more standard RL formulation of (state + action) -> reward

File: docs/source/hallway.rst (hunk 3 above)
Comment (davebiagioni) [reply docs/source/hallway.rst:228]: ok, thx for the feedback -- i erred on the side of putting too much detail here -- will remove

File: docs/source/hallway.rst (hunk 4 above)
Comment (davebiagioni) [reply docs/source/hallway.rst:362]: yeah exactly, will be sure to mention

**Comment by scottwsides** [comment]
I just wanted to double check where graph-env is building for you. I'm loading the needed dependencies and then trying to run the test_hallway.py test in a Chemistream container and getting a number of errors and messages about deprecated packages. I can send more details on slack, but I wanted to see first where you're getting this to run

File: graphenv/examples/hallway/hallway_state.py [hunk 5]
```diff
@@ -64,7 +64,7 @@ def reward(self) -> float:
             float: random reward between 0 and 2 on the goal vertex, -0.1
                 otherwise.
         """
-        return random.random() * 2 if self.cur_pos >= self.end_pos else -0.1
+        return random.random() * 2 if self.cur_pos >= self.end_pos - 1 else -0.1
```
Comment (pstjohn) [inline graphenv/examples/hallway/hallway_state.py:67]: weren't we planning to set `self.end_pos = corridor_length - 1`? That would probably be a simpler edit, although some of the tests might still need to get changed

File: graphenv/examples/hallway/hallway_state.py [hunk 6]
```diff
@@ -95,7 +95,7 @@ def _get_children(self) -> Sequence["HallwayState"]:
         Yields:
             HallwayState: Child verticies of this vertex.
         """
-        if self.cur_pos < self.end_pos:
+        if self.cur_pos < self.end_pos - 1:
```
Comment (pstjohn) [inline graphenv/examples/hallway/hallway_state.py:98]: likewise could revert this if we set `self.end_pos = corridor_length - 1`

**Review by pstjohn — COMMENTED** [review]
I'm happy to make those edits and change any tests to get the CI to pass

File: graphenv/examples/hallway/hallway_state.py (hunk 5 above)
Comment (davebiagioni) [reply graphenv/examples/hallway/hallway_state.py]: i like changing the end_pos definition, seems like a cleaner way to go -- pushed that change just now!
